clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -118,7 +118,12 @@ PVConstraint *ConstraintResolver::addAtom(PVConstraint *PVC, ConstAtom *PtrTyp,
 }
 
 // Processes E from malloc(E) to discern the pointer type this will be
-static ConstAtom *analyzeAllocExpr(Expr *E, Constraints &CS, QualType &ArgTy) {
+static ConstAtom *analyzeAllocExpr(CallExpr *CE, Constraints &CS, QualType &ArgTy, std::string FuncName) {
+  Expr *E;
+  if (!FuncName.compare("malloc"))
+    E = CE->getArg(0);
+  else
+    E = CE->getArg(1);
   ConstAtom *ret = CS.getPtr();
```

Comment:
I'm not sure this is the best way to handle `realloc`. Basically, you have something like this:
```
int *p = malloc(sizeof(int)*n);
...
int *q = realloc(p,sizeof(int)*m);
```
then basically there should be a flow from the first argument of `realloc` (here, `p`) to whatever is returned (here, `q`).  We can probably add extra constraints on the returned memory, to force an array in the case that the expression in the second argument has the right form. I think that's what you have. But you need the flow from `p` to `q` too (which is to say, the result of the `realloc` expression should include the `p` `ConstraintVariable` and perhaps a fresh one that should be constrained to be an array when needed.

Reply:
This makes sense to me. I'm a little confused design-wise on why we need p's constraint variable to flow into q. I thought the idea of a realloc was to completely invalidate whatever `p` is, since it becomes a dangling pointer after the call to realloc. Once a realloc call happens, the memory that p originally pointed to no longer matters, right?

Reply:
There's no mechanism in Checked C that stops you from reading a dangling pointer. This code is accepted by Checked C's compiler:
```
_Ptr<int> p = malloc(sizeof(int));
free(p);
*p = 1;
```
A call to `realloc` may or may not reallocate `p`. The man page says it will use the same space if possible, or else free it and allocate new space. The programmer should stop using `p`, but there's no enforcement of that.